lib/core/stream.nit

```diff
@@ -65,35 +95,71 @@ end
 abstract class Reader
 	super Stream
 
-	# Decoder used to transform input bytes to UTF-8
-	var decoder: Codec = utf8_codec is writable
+	# Read a byte directly from the underlying stream, without
+	# considering any eventual buffer
+	protected fun raw_read_byte: Int is abstract
+
+	# Read at most `max` bytes from the underlying stream into `buf`,
+	# without considering any eventual buffer
+	#
+	# Returns how many bytes were read
+	protected fun raw_read_bytes(buf: CString, max: Int): Int do
+		var rd = 0
+		for i in [0 .. max[ do
+			var b = raw_read_byte
+			if b < 0 then break
+			buf[i] = b.to_b
+			rd += 1
+		end
+		return rd
+	end
 
 	# Reads a character. Returns `null` on EOF or timeout
 	fun read_char: nullable Char is abstract
 
 	# Reads a byte. Returns a negative value on error
-	fun read_byte: Int is abstract
+	fun read_byte: Int do
+		var llen = lookahead_length
+		if llen == 0 then return raw_read_byte
+		var lk = lookahead
+		var b = lk[0].to_i
+		if llen == 1 then
+			lookahead_length = 0
+		else
+			lk.lshift(1, llen - 1, 1)
+			lookahead_length -= 1
+		end
+		return b
+	end
 
 	# Reads a String of at most `i` length
-	fun read(i: Int): String do return read_bytes(i).to_s
+	fun read(i: Int): String do
+		var cs = new CString(i)
+		var rd = read_bytes_to_cstring(cs, i)
+		return codec.decode_string(cs, rd)
+	end
 
-	# Read at most i bytes
-	#
-	# If i <= 0, an empty buffer will be returned
-	fun read_bytes(i: Int): Bytes
-	do
-		if last_error != null or i <= 0 then return new Bytes.empty
-		var s = new CString(i)
-		var buf = new Bytes(s, 0, i)
-		while i > 0 and not eof do
-			var c = read_byte
-			if c < 0 then
-				continue
-			end
-			buf.add c.to_b
-			i -= 1
+	# Reads up to `max` bytes from source
```

Comment:
When does it not read `max` chars? Only when at EOF or also when data is not yet available?

Reply:
That will depend on the implementation of `read_bytes_to_cstring`, I'd think it would be better to have a C-like semantic, i.e. reading up-to-`max` bytes if available, or the available buffered amount, while blocking on the first read if none are available.